Nested Executions initial implementation

rschnekenbu opened the pull request:
[WIP] Currently updating with new features implemented these last weeks (message reorient for example is not taken perfectly into account + anchoring for automatically created messages like reply message)

cdamus commented on plugins/org.eclipse.papyrus.uml.interaction.model/model/seqd.ecore:
```diff
@@ -487,6 +500,50 @@
+      <eParameters name="height" lowerBound="1" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EInt"/>
+      <eParameters name="metaclass" lowerBound="1" eType="ecore:EClass platform:/plugin/org.eclipse.emf.ecore/model/Ecore.ecore#//EClass"/>
+    </eOperations>
+    <eOperations name="elementAt" lowerBound="1">
```
We have an operation like this on the `MLifeline` class, and I can imagine we will need it in future on other classes like `MCombinedFragment` or `MInteractionOperand`.  Should we pull it up to `MElement`?  Perhaps raise an issue for that as a separate work item.

rschnekenbu replied:
Indeed. I have raised issue #397 to track this addition. I would prefer to avoid changing too much this current PR, as it is living for quite a long time ;)

cdamus commented on plugins/org.eclipse.papyrus.uml.interaction.model/src-gen/org/eclipse/papyrus/uml/interaction/internal/model/impl/MExecutionImpl.java:
```diff
@@ -147,6 +155,93 @@ public MLifeline getOwner() {
+		// browse all executions on lifeline after the start of this execution and before the end of this
+		// execution
+
+		List<MOccurrence<?>> occurences = getOwner().getOccurrenceSpecifications();
```
Typo:  should be 'occur**r**ences'.  Similarly for variables 'executionOccurences', 'occurence' below.

rschnekenbu replied:
fixed with commit 348a8d6

cdamus commented on plugins/org.eclipse.papyrus.uml.interaction.model/src-gen/org/eclipse/papyrus/uml/interaction/internal/model/impl/MLifelineImpl.java:
```diff
@@ -479,6 +483,57 @@ public Command nudgeHorizontally(int deltaX) {
+		List<MOccurrence<?>> occurrences = getOccurrenceSpecifications();
+
+		Stack<MExecution> stack = new Stack<>();
+		for (MOccurrence<?> occurence : occurrences) {
```
'occur**r**ence'

cdamus commented on plugins/org.eclipse.papyrus.uml.interaction.model/src-gen/org/eclipse/papyrus/uml/interaction/internal/model/impl/MLifelineImpl.java:
```diff
@@ -479,6 +483,57 @@ public Command nudgeHorizontally(int deltaX) {
+		// browse all executions on this lifeline and check the virtual owner
+		List<MOccurrence<?>> occurrences = getOccurrenceSpecifications();
+
+		Stack<MExecution> stack = new Stack<>();
```
We don't actually need a stack object.  We could just keep track of the depth of nesting with an `int`.

cdamus commented on plugins/org.eclipse.papyrus.uml.interaction.model/src-gen/org/eclipse/papyrus/uml/interaction/internal/model/impl/MLifelineImpl.java:
```diff
@@ -479,6 +483,57 @@ public Command nudgeHorizontally(int deltaX) {
+		List<MOccurrence<?>> orderedCoveredBys = getInteraction().getElement().getFragments().stream()
+				.filter(frg -> frg.getCovereds().contains(getElement())) //
+				.filter(OccurrenceSpecification.class::isInstance) //
+				.map(el -> getInteraction().getElement(el)) //
```
```suggestion
				.map(getInteraction()::getElement) //
```

cdamus commented on plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/commands/InsertNestedExecutionCommand.java:
```diff
@@ -0,0 +1,268 @@
+/**
+ * An execution specification creation operation.
+ *
+ * @author Christian W. Damus
```
Not me, no.  😉

rschnekenbu replied:
Oops ;-) Sorry, fixed in commit 348a8d6

cdamus commented on tests/org.eclipse.papyrus.uml.diagram.sequence.runtime.tests/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/policies/tests/ExecutionSpecificationCreationEditPolicyUITest.java:
```diff
@@ -0,0 +1,69 @@
+ */
+@SuppressWarnings("restriction")
+@ModelResource("one-exec.di")
+public class ExecutionSpecificationCreationEditPolicyUITest extends AbstractGraphicalEditPolicyUITest {
```
Nice! Unfortunately, this test fails on my Mac:

```
Expected: location point that x that 276 ± 1 and y that 170 ± 1
     but: location y was <165>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
	at org.eclipse.papyrus.uml.diagram.sequence.runtime.internal.edit.policies.tests.ExecutionSpecificationCreationEditPolicyUITest.createNestedExecution(ExecutionSpecificationCreationEditPolicyUITest.java:64)
```

On the plus side, all of the other tests still pass (with no new skips). 😉

rschnekenbu replied:
Hmmm, interesting! Again an issue with different platforms. I updated the test so it does get the initial Y value and compare the value after execution with the initial one. 
Funny fact is that this model is also used in other classes, like ExecutionSnappingUITest. In all other classes, it is not used directly, only with offset to set the message connection point, or not used at all ;)
I would still be very happy to understand this difference. The position of the execution is set on the lifeline body, at least in notation. The height of the header is set also, so as its y position. The interaction iself may have a different layout with its label on top, given the font? This is the only element that does not have a size pushed in notation.

cdamus replied:
Now, on my system, the tests are green.  So, that's good.  However, there's now an assumption failure that skips the `ExecutionMoveUITest::moveSyncMessageEnd` test.  It seems that the sync message is no longer located in the diagram where the test expects it to be.

cdamus commented on tests/org.eclipse.papyrus.uml.interaction.model.tests/src-gen/org/eclipse/papyrus/uml/interaction/model/tests/MLifelineTest.java:
```diff
@@ -733,20 +806,4 @@ public void testMakeCreatedAt__OptionalInt() {
 		assertThat("Center line not moved", getFixture().getTop(), isPresent(expectedTop.getAsInt()));
 	}
 
-	public void testMakeCreatedAt__OptionalInt__empty() {
```
How did this get deleted?  It didn't have an `@generated not` tag.  Please restore it.

rschnekenbu replied:
Restored in commit 348a8d6, probably a merge conflicts resolution failure on my side.

cdamus commented on tests/org.eclipse.papyrus.uml.interaction.model.tests/src/org/eclipse/papyrus/uml/interaction/model/tests/deletion/DeleteLifelineTest.java:
```diff
@@ -0,0 +1,170 @@
+
+@ModelResource({"DeleteLifeline.uml", "DeleteLifeline.notation" })
+@SuppressWarnings("boxing")
+public class DeleteLifelineTest {
```
This is a great addition to the tests!

rschnekenbu replied:
Thanks ;)

rschnekenbu commented on plugins/org.eclipse.papyrus.uml.diagram.sequence.runtime/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/policies/ExecutionSpecificationCreationEditPolicy.java:
```diff
@@ -0,0 +1,111 @@
+	protected Optional<org.eclipse.emf.common.command.Command> getCreationCommand(MInteraction interaction,
+			Element parentElement, View parentView, Point location, Dimension size, IElementType type) {
+
+		Optional<MExecution> mExecution = Optional.empty();
```
Thanks for the hint! That is fixed with commit 348a8d6

rschnekenbu commented on plugins/org.eclipse.papyrus.uml.diagram.sequence.figure/src/org/eclipse/papyrus/uml/diagram/sequence/figure/LifelineBodyFigure.java:
```diff
@@ -143,4 +137,15 @@ public void setDefaultAnchorDistance(int distance) {
 		this.defaultAnchorDistance = distance;
 	}
 
+	@SuppressWarnings("deprecation")
```
Unfortunately not, this code is the same as the overriden method. So it will be removed in an upcoming patch.

rschnekenbu commented:
I solved all the conversation items, I left the one about test failure on mac platform, because I am interested in understanding the diff here. For the other ones, commit 348a8d6 and e65b707 solves all of them.  

For the diagram behaviour:
- reorienting the message starting the nested execution moves now nicely the views. A reparentView was done while setting the owner of the executions, and so all moved elements views were added to the lifeline body figure. In the case of the nested execution, this is useless, as its parent is the nesting execution one. A check is now performed so it does not reparent views if not needed. A functional test was also added to check that layout is good on retargeting.
- For the nesting of executions, according to some live discussions with Antonio, it seems better to always cascade in the same direction, in case of big executions spanning across several computer screens. In this case, it would be easier to detect which one is the nesting one, and which are the nested ones if they are always cascading on the same side. So I propose to always keep the nesting on the right currently. We may update later if the users think this is not practical.
- for the side of the anchors, this is a bigger issue on executions in general. This will be handled in issue #376. I added a comment in that issue to ensure it is tested in the nested execution cases. I already started to implement some tests for that issue locally (currently failing of course!). This should be my next task.
- Good catch for the issue of creating an execution on connecting a sync message! I will check what is not going there, probably some edit policy missing on the execution similar to the one on lifeline body.
- For the test failures, I had fixed the inline comments. That should be fine now.

Finally, for the code formatting: looking at the history, it seems we have some trouble to keep consistency in formatting. I suspect we use different versions of EMF to generate code and / or do not apply the same actions after code generation. I did a format source code after generation / organize import and then a clean up using project profile (what I usually do when generating Papyrus diagram code). Even if I revert the edit plugin code and generate the code again, I get a lot of differences compared to the master branch.  I am using EMF 2.14 for code generation. 

Except for the automatic creation of the action execution when attaching a sync message to an execution, I addressed all the topics from your review.

Thank you again for that very valuable review!

cdamus commented:
> * reorienting the message starting the nested execution moves now nicely the views. A reparentView was done while setting the owner of the executions, and so all moved elements views were added to the lifeline body figure. In the case of the nested execution, this is useless, as its parent is the nesting execution one. A check is now performed so it does not reparent views if not needed. A functional test was also added to check that layout is good on retargeting.

Ah, that makes sense.  Good.

> * For the nesting of executions, according to some live discussions with Antonio, it seems better to always cascade in the same direction, in case of big executions spanning across several computer screens. In this case, it would be easier to detect which one is the nesting one, and which are the nested ones if they are always cascading on the same side. So I propose to always keep the nesting on the right currently. We may update later if the users think this is not practical.

Okay, I can understand that motivation.  Certainly, the worst would be to have nesting switch orientations within a stack, so this is a good compromise.

> * for the side of the anchors, this is a bigger issue on executions in general. This will be handled in issue #376. I added a comment in that issue to ensure it is tested in the nested execution cases. I already started to implement some tests for that issue locally (currently failing of course!). This should be my next task.

Cool.  TDD FTW 😉 

> * Good catch for the issue of creating an execution on connecting a sync message! I will check what is not going there, probably some edit policy missing on the execution similar to the one on lifeline body.

Yes, that sounds likely.  Another good motivation for moving more of this logic out of the edit-policies and into the logical model where it can be shared by the edit-policies.

> Finally, for the code formatting: looking at the history, it seems we have some trouble to keep consistency in formatting. I suspect we use different versions of EMF to generate code and / or do not apply the same actions after code generation. I did a format source code after generation / organize import and then a clean up using project profile (what I usually do when generating Papyrus diagram code). Even if I revert the edit plugin code and generate the code again, I get a lot of differences compared to the master branch. I am using EMF 2.14 for code generation.

Yeah, I'm using Eclipse for Committers 2018-09, so that's EMF 2.15.  Those templates can have some changes, hopefully only improvements, because the genmodel is still targeting EMF 2.14 run-time.  However, our project configurations for JDT formatting and clean-up should take care of purely formatting issues (though the generated code may actually be semantically different).

_Probably we should standardize on the development environment, perhaps via an Oomph setup model for this project._

rschnekenbu commented on tests/org.eclipse.papyrus.uml.diagram.sequence.runtime.tests/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/policies/tests/ExecutionMoveUITest.java:
```diff
@@ -0,0 +1,108 @@
+	public void moveSyncMessageEnd() {
+		/* initial state - check message end position */
+		EditPart syncEP = requireMessage("nested-execution::interaction1::sync");
+		assumeThat(syncEP, runs(LIFELINE_2__BODY_X, SYNC_Y, LIFELINE_3__BODY_X - (EXEC_WIDTH / 2), SYNC_Y));
```
This one is fixed with commit 4ac2cc3

rschnekenbu commented:
Solving conflicts by merging the master branch. This commit should be hopefully the last one on this PR!
